### PR TITLE
add test cases to invalidateFramebuffer and invalidateSubFramebuffer against WebGL 2

### PR DESCRIPTION
**Pull request by Richard-Yunchao**
PTAL when you have time.

invalidate framebuffer on default framebuffer in chromium will fail. I suppose there are some errors in command buffer in chromium.

### CODE REVIEW

File: sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            'should be no errors after bliting framebuffer.');
+ 
+  // invalidate the multi-sampled color attachment of the bound read framebuffer after blitFramebuffer.
+  invalidation(gl.READ_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.DEPTH_ATTACHMENT);
+
+  var maxColorAttachments = gl.getParameter(gl.MAX_COLOR_ATTACHMENTS);
+  gl.invalidateSubFramebuffer(gl.READ_FRAMEBUFFER, [gl.COLOR_ATTACHMENT0 + maxColorAttachments], 5, 5, 10, 10);
+  wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION,
+            "calling invalidateSubFramebuffer to invalidate a COLOR_ATTACHMENT that exceeds MAX_COLOR_ATTACHMENT should generate INVALID_OPERATION.");
+  gl.invalidateFramebuffer(gl.READ_FRAMEBUFFER, [gl.COLOR_ATTACHMENT0 + maxColorAttachments], 5, 5, 10, 10);
```
Comment (zhenyao) [inline sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html]: invalidateFramebuffer() with too many arguments.

File: sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  gl.deleteRenderbuffer(rb);
+  gl.deleteFramebuffer(fb);
+  gl.deleteRenderbuffer(buffer);
+}
+
+function invalidation(target, valid_attachment, invalid_attachment) {
+  gl.invalidateSubFramebuffer(target, [invalid_attachment], 5, 5, 10, 10);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling invalidateSubFramebuffer to invalidate a specified attachment which is not existed will be ignored. This should succeed.");
+  gl.invalidateFramebuffer(target, [valid_attachment], 5, 5, 10, 10);
```
Comment (zhenyao) [inline sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html]: Too many arguments

File: sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling invalidateSubFramebuffer should succeed.");
+
+  gl.invalidateFramebuffer(target, [valid_attachment], -5, -5, 10, 10);
```
Comment (zhenyao) [inline sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html]: Too many arguments

File: sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling invalidateSubFramebuffer should succeed, even the invalidated pixels may be outside of the framebuffer allocated to current context. These pixels are ignored.");
+  gl.invalidateFramebuffer(target, [valid_attachment], 5, 5, 20, 20);
```
Comment (zhenyao) [inline sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html]: Too many arguments

**Comment by Richard-Yunchao** [comment]
@zhenyao , Sorry, I uploaded a wrong branch from my local desktop, :( 

Yeah, these 'invalidateFramebuffer' should be 'invalidateSubFramebuffer'. Code has been updated, PTAL.

**Comment by kenrussell** [comment]
What's the status of this test now? Does it pass?

File: sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  var fb_m = gl.createFramebuffer();
+  gl.bindFramebuffer(gl.FRAMEBUFFER, fb_m);
+
+  var rb_m = gl.createRenderbuffer();
+  gl.bindRenderbuffer(gl.RENDERBUFFER, rb_m);
+  var samples = gl.getInternalformatParameter(gl.RENDERBUFFER, gl.RGBA8, gl.SAMPLES);
+  gl.renderbufferStorageMultisample(gl.RENDERBUFFER, samples[0], gl.RGBA8, canvas.width, canvas.height);
+  gl.framebufferRenderbuffer(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.RENDERBUFFER, rb_m);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "should be no errors after attaching a multi-sampled renderbuffer to fbo.");
```
Comment (zhenyao) [inline sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html]: weird indent, here and below

File: sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+  var maxColorAttachments = gl.getParameter(gl.MAX_COLOR_ATTACHMENTS);
+  gl.invalidateSubFramebuffer(gl.READ_FRAMEBUFFER, [gl.COLOR_ATTACHMENT0 + maxColorAttachments], 5, 5, 10, 10);
+  wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION,
+            "calling invalidateSubFramebuffer to invalidate a COLOR_ATTACHMENT that exceeds MAX_COLOR_ATTACHMENT should generate INVALID_OPERATION.");
+  gl.invalidateFramebuffer(gl.READ_FRAMEBUFFER, [gl.COLOR_ATTACHMENT0 + maxColorAttachments]);
+  wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION,
+            "calling invalidateFramebuffer to invalidate a COLOR_ATTACHMENT that exceeds MAX_COLOR_ATTACHMENT should generate INVALID_OPERATION.");
+
+  gl.bindFramebuffer(gl.FRAMEBUFFER, fb);
```
Comment (zhenyao) [inline sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html]: This is unnecesary

File: sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  gl.deleteRenderbuffer(rb_m);
+  gl.deleteRenderbuffer(rb);
+  gl.deleteFramebuffer(fb);
+  gl.deleteRenderbuffer(buffer);
+}
+
+function invalidation(target, valid_attachment, invalid_attachment) {
+  gl.invalidateSubFramebuffer(target, [invalid_attachment], 5, 5, 10, 10);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling invalidateSubFramebuffer to invalidate a specified attachment which is not existed will be ignored. This should succeed.");
```
Comment (zhenyao) [inline sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html]: "which is not existed" -> "that does not exist"

File: sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+  gl.invalidateSubFramebuffer(target, [valid_attachment], -5, -5, 10, 10);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling invalidateSubFramebuffer should succeed, even the invalidated pixels may be outside of the framebuffer allocated to current context. These pixels are ignored.");
+  gl.invalidateSubFramebuffer(target, [valid_attachment], 5, 5, 20, 20);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling invalidateSubFramebuffer should succeed, even the invalidated pixels may be outside of the framebuffer allocated to current context. These pixels are ignored.");
+
+  gl.invalidateFramebuffer(target, [invalid_attachment]);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling invalidateFramebuffer to invalidate a specified attachment which is not existed will be ignored. This should succeed.");
```
Comment (zhenyao) [inline sdk/tests/conformance2/renderbuffers/invalidate-framebuffer.html]: "which is not existed" -> "that does not exist"

**Comment by zhenyao** [comment]
Mostly look good with a few nits.  There is one case you did not test, if the framebuffer is not complete (for example, no images are attached), calling invalidate{Sub}Framebuffer() will be no op and no error.

Could you also add a test for that? So we have all cases covered.

**Comment by Richard-Yunchao** [comment]
@zhenyao , thanks a lot for your reviewing, code has been updated accordingly.  @kenrussell , invalidate{Sub}Framebuffer() failed on default framebuffer in chromium.

**Comment by zhenyao** [comment]
Looks good.  Thank you.

**Pull request merged**